File: C++/Linked List/02_Reverse_with_group_size.cpp
/*
Idea:
Using dummy variable to keep track of the head after being sorted.
& reversing the linked list for each group.
*/

//{ Driver Code Starts
#include<bits/stdc++.h>
using namespace std;


struct node
{
    int data;
    struct node* next;

    node(int x) {
        data = x;
        next = NULL;
    }

};

/* Function to print linked list */
void printList(struct node* node)
{
    while (node != NULL)
    {
        printf("%d ", node->data);
        node = node->next;
    }
    printf("\n");
}


class Solution
{
public:
    struct node* reverse(struct node* head, int k)
    {
        if (!head || k == 1)
            return head;
        node* dummy = new node();
        dummy->data = -1;
        dummy->next = head;

        int length = 0;
        node* prev = dummy, * next = dummy, * now = dummy;
        while (now != NULL)
        {
            now = now->next; // finding length of the linked list
            length++;
        }

        while (next)
        {
            now = prev->next;
            next = now->next; // setting up vars for each group

            int tempLen = length > k ? k : length - 1; // finding the group length is of target length of 
            for (int i = 1;i < tempLen;++i)
            {
                now->next = next->next;
                next->next = prev->next;
                prev->next = next;
                next = now->next;
                // reversing the group 
            }
            prev = now;
            length -= k;
        }
        return dummy->next; // returning the head
    }
};


//{ Driver Code Starts.

/* Drier program to test above function*/
int main(void)
{
    int t;
    cin >> t;

    while (t--)
    {
        struct node* head = NULL;
        struct node* temp = NULL;
        int n;
        cin >> n;

        for (int i = 0; i < n; i++)
        {
            int value;
            cin >> value;
            if (i == 0)
            {
                head = new node(value);
                temp = head;
            }
            else
            {
                temp->next = new node(value);
                temp = temp->next;
            }
        }

        int k;
        cin >> k;

        Solution ob;
        head = ob.reverse(head, k);
        printList(head);
    }

    return(0);
}


// } Driver Code Ends

File: Python/Linked List/02_Reverse_with_group_size.py
"""
Idea:
Using dummy variable to keep track of the head after being sorted.
& reversing the linked list for each group.
"""


class Solution:
    def reverse(self, head, k):
        if not head or k == 1:
            return head

        dummy = Node(-1)
        dummy.next = head

        length = 0
        prev = dummy
        next_node = dummy
        now = dummy

        while now is not None:
            now = now.next  # Finding the length of the linked list
            length += 1

        while next_node:
            now = prev.next
            next_node = now.next  # Setting up variables for each group

            temp_len = min(length-1, k)  # Finding the group length, limited to the target length of k
            for i in range(1, temp_len):
                now.next = next_node.next
                next_node.next = prev.next
                prev.next = next_node
                next_node = now.next  # Reversing the group

            prev = now
            length -= k

        return dummy.next  # Returning the head


# {
# Driver Code Starts
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class LinkedList:
    def __init__(self):
        self.head = None
        # self.tail

    def push(self, new_data):
        new_node = Node(new_data)
        new_node.next = self.head
        self.head = new_node

    def printList(self):
        temp = self.head
        while (temp):
            print(temp.data, end=" ")
            # arr.append(str(temp.data))
            temp = temp.next
        print()


if __name__ == '__main__':
    t = int(input())
    while (t > 0):
        llist = LinkedList()
        n = input()
        values = list(map(int, input().split()))
        for i in reversed(values):
            llist.push(i)
        k = int(input())
        new_head = LinkedList()
        ob = Solution()
        new_head = ob.reverse(llist.head, k)
        llist.head = new_head
        llist.printList()
        t -= 1

# } Driver Code Ends
